Add CDI support for additional JPA/Hibernate components

ayagmar opened the pull request:
Adding CDI support for the following JPA/hibernate components:

- CriteriaBuilder
- Metamodel
- Cache
- PersistenceUnitUtil
- SchemaManager

**Testing**

In addition to unit tests, I verified my changes in a quarkus project using local build to ensure proper injection and runtime behavior 
![{F2BB58DD-8CEB-432C-9C2D-33566A56E488}](https://github.com/user-attachments/assets/f99c9615-b75f-4ae0-9c9d-a30c7c391238)


**q/arc/beans**
```
  {
    "id": "4skIeF8wp4Wg4D6tne7GHU-9epc",
    "kind": "SYNTHETIC",
    "generatedClass": "org.hibernate.query.criteria.HibernateCriteriaBuilder_4skIeF8wp4Wg4D6tne7GHU-9epc_Synthetic_Bean",
    "beanClass": "org.hibernate.query.criteria.HibernateCriteriaBuilder",
    "types": [
      "jakarta.persistence.criteria.CriteriaBuilder",
      "org.hibernate.query.criteria.HibernateCriteriaBuilder",
      "java.lang.Object"
    ],
    "qualifiers": [
      "@Default",
      "@Any"
    ],
    "scope": "jakarta.enterprise.context.ApplicationScoped"
  }
```

Please let me know if you'd like any refinements or further adjustments. 
Thanks!

- Resolves: #46843 

FYI : @yrodiere @gsmet

ayagmar committed:
Add CDI support for CriteriaBuilder, Metamodel, Cache, and PersistenceUnitUtil in Hibernate ORM (#46843)

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmCdiProcessor.java:
```diff
@@ -251,6 +302,47 @@ void generateDataSourceBeans(HibernateOrmRecorder recorder,
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            CriteriaBuilder.class, CRITERIA_BUILDER_EXPOSED_TYPES, false)
+                            .createWith(recorder.criteriaBuilderSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            jakarta.persistence.metamodel.Metamodel.class, METAMODEL_EXPOSED_TYPES, false)
+                            .createWith(recorder.metamodelSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            SchemaManager.class, SCHEMA_MANAGER_EXPOSED_TYPES, false)
+                            .createWith(recorder.schemaManagerSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            Cache.class, CACHE_EXPOSED_TYPES, false)
+                            .createWith(recorder.cacheSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            jakarta.persistence.PersistenceUnitUtil.class, PERSISTENCE_UNIT_UTIL_EXPOSED_TYPES, false)
+                            .createWith(recorder.persistenceUnitUtilSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
```
Given this is becoming a lot of duplicated code, maybe you could extract it to a separate method (along with the Session/StatelessSession definition) and call the method both here and above?

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmCdiProcessor.java:
```diff
@@ -206,6 +216,47 @@ void generateDataSourceBeans(HibernateOrmRecorder recorder,
+                            true, true,
+                            CriteriaBuilder.class, CRITERIA_BUILDER_EXPOSED_TYPES, false)
+                            .createWith(recorder.criteriaBuilderSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
```
Not 100% necessary, but it would be a good idea to inject the `SessionFactory` instead. That will lead to better error messages when trying to inject the SchemaManager of an inactive persistence unit, once we implement #43594.

Yes I know, you asked that question, but I was offline, sorry :)

ayagmar replied:
Hey I will try to do this hopefully nothing unexpected pops out haha and no worries

yrodiere reviewed (changes requested):
Thanks a lot, this is awesome work, especially the testing.

I added a few comments, but only the exposed types are really important -- we could possibly address the rest as follow-ups. Let me know if you prefer that!

ayagmar commented:
> Thanks a lot, this is awesome work, especially the testing.
> 
> I added a few comments, but only the exposed types are really important -- we could possibly address the rest as follow-ups. Let me know if you prefer that!

Hey thanks for the review! I’ll work on addressing all the comments now and especially exposed the other types since they extend it.. 
I’ll update the PR once everything is in a good, reviewable state and nothing’s broken

ayagmar commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/ClassNames.java:
```diff
@@ -53,6 +53,11 @@ private static DotName createConstant(String fqcn) {
     public static final DotName SESSION = createConstant("org.hibernate.Session");
     public static final DotName STATELESS_SESSION = createConstant("org.hibernate.StatelessSession");
+    public static final DotName CRITERIA_BUILDER = createConstant("jakarta.persistence.criteria.CriteriaBuilder");
+    public static final DotName METAMODEL = createConstant("jakarta.persistence.metamodel.Metamodel");
```
done

yrodiere commented on .gitignore:
```diff
@@ -31,9 +28,8 @@ ObjectStore
-quarkus.log*
-replay_*.logß
+*.log
+*.log*
```
@gsmet Are you okay with this? Seems fine to me, but since it impacts more than my little part of the project...

gsmet replied:
Yeah, I had seen it and I think we can try this approach and see if it becomes annoying.

yrodiere replied:
Taken offline, Guillaume is okay to try. Let's!

yrodiere commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java:
```diff
@@ -146,6 +154,59 @@ public StatelessSession get() {
+    private SessionFactory getSessionFactoryReference(SyntheticCreationalContext<?> context, String persistenceUnitName) {
+        Class<SessionFactory> beanType = SessionFactory.class;
+        if (PersistenceUnitUtil.isDefaultPersistenceUnit(persistenceUnitName)) {
+            InjectableBean<?> defaultBean = Arc.container().instance(beanType, Default.Literal.INSTANCE).getBean();
+            if (defaultBean == null) {
+                return null;
+            }
```
Can this actually happen? If you didn't witness it, I'd leave the code out.

ayagmar replied:
I didn't witness it it was more of a defensive check
CDI will throw an exception in any case if something is off so I have removed it
Thanks!

yrodiere commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java:
```diff
@@ -156,11 +217,8 @@ public void doValidation(HibernateOrmRuntimeConfig hibernateOrmRuntimeConfig, St
         if (!"none".equals(schemaManagementStrategy)) {
             return;
         }
-        new Thread(new Runnable() {
-            @Override
-            public void run() {
-                SchemaManagementIntegrator.runPostBootValidation(puName);
-            }
-        }, "Hibernate post-boot validation thread for " + puName).start();
+        new Thread(
+                () -> SchemaManagementIntegrator.runPostBootValidation(puName),
+                "Hibernate post-boot validation thread for " + puName).start();
```
I've been told repeatedly to avoid lambdas in Quarkus extension code, for weird reasons related to memory/performance.

I won't ask that you avoid them everywhere, but please don't go out of your way to introduce more lambdas. Please revert this particular change? :)

yrodiere reviewed (commented):
Just a doubt on the CDI bean retrieval, and an unrelated change that would be nice to revert. Otherwise looks good!

yrodiere commented:
LGTM, thanks!
